mission/mission_handler_takeoff.hxx

```diff
@@ -0,0 +1,151 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+/*******************************************************************************
+ * \file mission_handler_takeoff.hxx
+ *
+ * \author MAV'RIC Team
+ * \author Matthew Douglas
+ *
+ * \brief The MAVLink mission planner handler functions for the takeoff state
+ *
+ ******************************************************************************/
+
+
+#ifndef MISSION_HANDLER_TAKEOFF_HXX__
+#define MISSION_HANDLER_TAKEOFF_HXX__
+
+template <class T>
+Mission_handler_takeoff<T>::Mission_handler_takeoff(T& controller,
+                                                    const INS& ins,
+                                                    State& state):
+            Mission_handler(),
+            controller_(controller),
+            ins_(ins),
+            state_(state)
+{
+    waypoint_ = Waypoint(   MAV_FRAME_LOCAL_NED,
+                            MAV_CMD_NAV_TAKEOFF,
+                            0,
+                            0.0f,
+                            0.0f,
+                            0.0f,
+                            0.0f,
+                            0.0f,
+                            0.0f,
+                            0.0f);
+}
+
+template <class T>
+bool Mission_handler_takeoff<T>::can_handle(const Waypoint& wpt) const
+{
+    bool handleable = false;
+
+    uint16_t cmd = wpt.command();
+    if (cmd == MAV_CMD_NAV_TAKEOFF)
+    {
+        handleable = true;
+    }
+
+    return handleable;
+}
+
+template <class T>
+bool Mission_handler_takeoff<T>::setup(const Waypoint& wpt)
+{
+    bool success = true;
+
+    waypoint_ = wpt;
+
+    print_util_dbg_print("Automatic take-off, will hold position at: (");
+    print_util_dbg_print_num(wpt.local_pos()[X], 10);
+    print_util_dbg_print(", ");
+    print_util_dbg_print_num(wpt.local_pos()[Y], 10);
+    print_util_dbg_print(", ");
+    print_util_dbg_print_num(wpt.local_pos()[Z], 10);
+    print_util_dbg_print(")\r\n");
+
+    return success;
+}
+
+template <class T>
+Mission_handler::update_status_t Mission_handler_takeoff<T>::update()
+{
+    // Set goal
+    bool ret = set_control_command();
+
+    /*******************
+    Determine status code 
+    ********************/
+    bool finished = false;
+    
+
+    // Determine distance to the waypoint
+    local_position_t wpt_pos = waypoint_.local_pos();
+    float xy_radius_sqr = wpt_pos[Z]*wpt_pos[Z]*0.16f;
+
+    float xy_dist2wp_sqr;
+    float rel_pos[3];
+    for (int i = 0; i < 2; i++)
+    {
+        rel_pos[i] = wpt_pos[i] - ins_.position_lf()[i];
+    }
+    rel_pos[2] = 0.0f;
+    xy_dist2wp_sqr = vectors_norm_sqr(rel_pos);
+
+    // Determine if finished
+    if (xy_dist2wp_sqr <= xy_radius_sqr && ins_.position_lf()[Z] <= 0.9f * wpt_pos[Z])
+    {
+        finished = true;
+    }
+
+    // Return true if waiting at waypoint and autocontinue
+    if (finished && waypoint_.autocontinue() == 1)
```

Comment:
If we check the autocontinue in Mission_planner, it would reduce some flexibility. For example, if we want to ignore the autocontinue flag for some reason (like a timeout if no commands have been received within a set time), the handler could advance anyways. 

Though as you said, autocontinue checks within the mission planner would make it less error prone.